### PR TITLE
fix asm2wasm import type setting - set the type of used calls based o…

### PR DESCRIPTION
**Pull request by kripken**
…n the context, early, so it's valid in the optimizer

(see comment in source for details)

### CODE REVIEW

File: src/asm2wasm.h [hunk 1]
```diff
@@ -1980,9 +1979,17 @@ Function* Asm2WasmBuilder::processFunction(Ref ast) {
...
+          // import which does return a value, and we use that elsewhere. finalizeCalls
+          // fixes that up. what we do here is wherever a value is used, we set the right
+          // value, which is enough to ensure that the wasm ast is valid for such uses.
+          // this is important as we run the optimizer on functions before we get
+          // to finalizeCalls (which we can only do once we've read all the functions,
+          // and we optimize in parallel starting earlier).
           Ref parent = astStackHelper.getParent();
-          WasmType type = !!parent ? detectWasmType(parent, &asmData) : none;
-          noteImportedFunctionCall(ast, type, ret->cast<CallImport>());
+          import->type = !!parent ? detectWasmType(parent, &asmData) : none;
```
Comment (jgravelle-google) [inline src/asm2wasm.h]: My first thought: should this be wrapped with `if (import->type == none)` ?
Reading more closely, `import` is the call site, so this doesn't overwrite anything else.
Would it make sense to rename `import` to `callImport` to make it clearer?

File: src/asm2wasm.h [hunk 2]
```diff
@@ -1980,9 +1979,17 @@ Function* Asm2WasmBuilder::processFunction(Ref ast) {
...
+          noteImportedFunctionCall(ast, import->type, ret->cast<CallImport>());
```
Comment (jgravelle-google) [inline src/asm2wasm.h]: Instead of `ret->cast<CallImport>()`, how about just `import`?

File: src/asm2wasm.h (hunk 1 above)
Comment (kripken) [inline src/asm2wasm.h]: Good idea, fixing.

File: src/asm2wasm.h (hunk 2 above)
Comment (kripken) [inline src/asm2wasm.h]: Also good idea, thanks.

**Comment by jgravelle-google** [comment]
lgtm

**Pull request merged**